src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -347,6 +347,10 @@ private static Binder GetEnclosingBinderInternalWithinRoot(SyntaxNode node, int
                         binder = rootBinder.GetBinder(current);
                     }
                 }
+                else if (current is RecordDeclarationSyntax { ParameterList: not null })
+                {
+                    binder = rootBinder.GetBinder(current);
```

Comment:
>binder = rootBinder.GetBinder(current); [](http://example.com/codeflow?start=20&length=39)

This doesn't feel right. It looks like this is going to bring locals declared in ```PrimaryConstructorBaseType``` in scope at position outside of that node, which would be incorrect (```PrimaryConstructorBaseType``` is handed above). Normally problems like this can be observed though LookupNames/Symbols APIs. However, it is quite possible that the way we decide which position is handled by a member semantic model prevents the exposure. Even if that is the case, and unless I am missing something, I think we can benefit from a clean up here. In particular, it looks like we don't need to create an ```ExpressionVariableBinder``` for a ```RecordDeclarationSyntax``` in ```LocalBinderFactory```. Instead we can create it for ```PrimaryConstructorBaseTypeSyntax``` and make necessary adjustments in ```Binder.BindRecordConstructorBody```.

Reply:
> It looks like this is going to bring locals declared in PrimaryConstructorBaseType in scope at position outside of that node, which would be incorrect (PrimaryConstructorBaseType is handed above).

I'm not sure this is true. Locals declared in the `PrimaryConstructorBaseType` will be scoped to the binder that is created in the binder factory for that syntax location. I can add a test to verify that this is the case if you want: I verified this locally and tested that an out variable declared in the base type constructor call was not visible in `LookupSymbols` in the attribute or in the body of the record class declaration.

Reply:
Added `AttributedDerivedRecord_BaseParameterNotVisibleInBody`.

Reply:
>>It looks like this is going to bring locals declared in PrimaryConstructorBaseType in scope at position outside of that node, which would be incorrect (PrimaryConstructorBaseType is handed above).

>I'm not sure this is true.

This is easy to observe under debugger:
![image](https://user-images.githubusercontent.com/10522107/194094406-ed64e92f-f9c8-4f1e-8826-47f7b6107a1e.png)

Note, that ```AttributedDerivedRecord_BaseParameterNotVisibleInBody``` hits this code path only once under "umbrella" of ```EnsureNullabilityAnalysisPerformedIfNecessary``` call, which doesn't affect symbols found by either ```LookupSymbols``` call in the test. As I mentioned in the original comment, it might be tricky, or perhaps even impossible to expose the problem outside due to the way we lookup member semantic models. This, however, doesn't mean that what the code is doing is the right thing to do.

> Locals declared in the ```PrimaryConstructorBaseType``` will be scoped to the binder that is created in the binder factory for that syntax location.

I agree that this is how it should be. However, as I mentioned in my initial comment, that binder is created for the ```RecordDeclarationSyntax``` instead, which causes the assert here. And I suggested the way to address that. We can discuss this offline in more details.